python_modules/dagster/dagster/core/asset_defs/decorators.py

```diff
@@ -168,6 +170,11 @@ def __call__(self, fn: Callable) -> AssetsDefinition:
             else None,
         )
 
+        # temporary workaround to retrieve asset definition from job
+        metadata["asset_def"] = asset_def
```

Comment:
This feels a little sketchy to me.  I might prefer having this on `Out` / `OutputDefinition` as some experimental argument than putting it in to the metadata.

Sidenote:  I think this works by mutating the metadata after it's passed to the `Out`, which is a little risky.  Might be safer to first define the asset_def, then pass it into the out.

Reply:
If we added an experimental argument to `OutputDefinition`, my only concern is that users may input some argument when that isn't the intended use case--is this possible? If not, I can do the second option of passing the metadata into the out after defining the asset.

Reply:
This doesn't seem that big of a concern, especially since they'll have to manually construct an asset definition (since we can actually do type checks against it), unlike this `metadata` object